Improve README/CONTRIBUTING [WIP] (#666)
zeit/docs: CONTRIBUTING.md

# Contributing to the ZEIT Documentation

Thank you for your interest in contributing to the ZEIT Docs!

> The following describes how to contribute to the ZEIT documentation, API reference, examples list, or guides website with either code or content.

Reviewer: ```suggestion
The following describes how to contribute to the ZEIT documentation, API reference, guides and code examples.
```

## What Does this Document Contain?

This document contains all the required information and links to resources needed to contribute to the ZEIT documentation:

- [Code of Conduct](#code-of-conduct) - our Code of Conduct, adapted from the [Contributor Covenant](http://contributor-covenant.org)

> - [Reporting an Issue](#reporting-issues) - guidance on how to raise an issue or provide feedback on the Zeit documentation

Reviewer: ```suggestion
- [Reporting an Issue](#reporting-issues) - guidance on how to raise an issue or provide feedback on the ZEIT documentation
```

- [Contributing Content](#contributing-content) - a collection of all information required for contributing

>   - [Forking Workflow](#forking-workflow) - a guide on how to keep your fork in sync with the official repository

Reviewer: ```suggestion
  - [Forking Workflow](#forking-workflow) - a guide on how to keep your fork in sync with `master`
```

`git pull upstream master`

This will get the most recent changes and bring your local fork up to date, at this point it is a good idea to have your remote fork reflect this, all this requires is to use:

`git push`

Now, both your local and remote fork are up to date with the official repository, you are ready to create branches and make changes!

### Pull Request Changes Requested

When you submit a pull request, a reviewer may request that you make further changes before merging with the master branch. This situation might seem difficult at first but is extremely straightforward.

All you need to do is make the changes on the same branch locally, commit and push them. Because you already have a pull request set up to track that branch, GitHub will automatically update it with your most recent changes. At this point it's a good idea to tag the original reviewer in the pull request using `@<reviewers username>`, to let them know you've made the requested changes.

## Platform and Usage Documentation

> The content for this section can be found in the `pages/docs` directory under version tags. We are currently focused on developing our documentation for Now platform version 2 and so the content for this version can be found within the `pages/docs/v2` directory.

Reviewer: ```suggestion
The content for this section can be found in the `pages/docs` directory, under its respective version tag. The latest documentation for Now 2.0 can be found within the `pages/docs/v2` directory.
```

As suggested by the `.mdx` extension, the ZEIT documentation uses [MDX](https://mdxjs.com), a markdown parser mixed with JSX. This allows us to use React components within markdown to nicely render our content.

An example of a docs MDX file:

```jsx
import Doc from '~/components/layout/docs'

import { TerminalInput } from '~/components/text/terminal'

export const meta = {
  title: 'The Title for the New Doc',
  description: 'The description for the new documentation page.',
  date: '1 January 2019'
}

This is the content written in Markdown with MDX!.

<TerminalInput># this is how we show the terminal input</TerminalInput>

The following is to allow the content to be exported as a page with our layout.

export default ({ children }) => <Doc meta={meta}>{children}</Doc>
```

When adding a new page to the v2 docs, make sure to add it to the appropriate section with a URL to the object located in the `lib/data/v2/docs.js` file.

## Guides

The content for this section can be found in the `pages/guides` directory.

> It uses the same method as [Platform and Usage Docs](#platform-and-usage-documentation) for content, using MDX for files and those files being top-level pages.

Reviewer: ```suggestion
Just like [Platform and Usage Docs](#platform-and-usage-documentation), our Guides use the same MDX system to generate top-level pages.
```

A guide can be written on many topics and will commonly focus on the creation and deployment of a project using Now. We strongly encourage you to take a look at the existing guides, these illustrate the right level of detail required for each step, from initial creation through to deployment.

> Guides are not limited to project creation and deployment however. The Now platform is incredibly powerful and feature-rich, as a result there are many topics you could contribute a guide on, our advice would be to take a feature you enjoy and use that as a starting point for your guide.

Reviewer: ```suggestion
Guides are not limited to project creation and deployment however. The Now platform is incredibly powerful and feature-rich, as a result there are many topics you could contribute to, our advice would be to take a feature you enjoy and use that as a starting point for your guide.
```

When adding a new guide, make sure to import and export the meta of the guide in the `lib/data/guides.js` file. This enables the guide to be listed on the guides front-page.

## API Reference

The content for this section can be found in the `pages/docs/api` directory. It uses the same versioning structure as `pages/docs`.

We are currently focused on expanding the documentation for the Now platform 2.0 API reference which can be found in `pages/docs/api/v2`.

As with the [Platform and Usage Docs](#platform-and-usage-documentation), the API uses MDX, although each file is not an independent page but instead is compiled into a single page, listing the file in the `lib/data/v2/api.js` file.

An example API Reference MDX file: https://github.com/zeit/docs/blob/master/pages/docs/api/v2/api-docs-mdx/api-basics/rate-limits.mdx

## Examples

The content for the examples section is not contained within this repository. To contribute an example, please head over to the [now-examples](https://github.com/zeit/now-examples) repository where you will find all the relevant information to do so.

## Using Components

> To make the creation of documentation and guides as simple as possible, we have developed a [component library](https://github.com/zeit/docs/tree/master/components). This library allows us to keep the format and appearance of our documentation and guides consistent, it also means that putting new content together doesn't take too long!

Reviewer: ```suggestion
To make the creation of documentation and guides as simple as possible, we have developed a [component library](https://github.com/zeit/docs/tree/master/components). This library allows us to keep the format and appearance of our documentation and guides consistent, making creating new content a breeze!
```

You can find multiple examples in each page of how to import and use these components, below is an example import statement:

`import { GenericLink } from '~/components/text/link'`

We believe that these components should be enough to cover usage in nearly all situations, however, if you feel that a new component should be added then please do so in a separate pull request, stating clearly your reasons for adding it.

## Static Assets

> A picture says a thousand words! Using images in guides and documentation can really help the reader understand the point you are putting across, if you'd like to include images, or any other static asset, you should follow this advice.

Reviewer: ```suggestion
A picture is worth a thousand words. Using graphics such as images in guides and documentation can really help the reader understand new concepts, if you'd like to include images, or any other static asset, you should follow this advice.
```
